nixops/backends/digital_ocean.py

```diff
@@ -0,0 +1,169 @@
+    size = nixops.util.attr_property("digital-ocean.size", None)
+    auth_token = nixops.util.attr_property("digital-ocean.authToken", None)
+    droplet_id = nixops.util.attr_property("digital-ocean.dropletId", None)
+    key_pair = nixops.util.attr_property("digital-ocean.keyPair", None)
+
+    def __init__(self, depl, name, id):
+        MachineState.__init__(self, depl, name, id)
+        self.name = name
+
+    def get_ssh_name(self):
+        return self.public_ipv4
+
+    def get_ssh_flags(self, *args, **kwargs):
+        super_flags = super(DigitalOceanState, self).get_ssh_flags(*args, **kwargs)
+        return super_flags + [
+            '-o', 'UserKnownHostsFile=/dev/null',
+            '-o', 'StrictHostKeyChecking=no',
+            '-i', self.get_ssh_private_key_file(),
+        ]
+
+    def get_physical_spec(self):
+        prefixLength = bin(int(socket.inet_aton(self.netmask).encode('hex'), 16)).count('1')
+
+        return Function("{ ... }", {
+            'imports': [ RawValue('<nixpkgs/nixos/modules/profiles/qemu-guest.nix>') ],
+            'networking': {
+                'defaultGateway': self.default_gateway,
+                'nameservers': ['8.8.8.8'], # default provided by DO
+                ('interfaces', 'enp0s3'): {
+                    'ip4': [{"address": self.public_ipv4, 'prefixLength': prefixLength}],
+                },
+            },
+            ('boot', 'loader', 'grub', 'device'): 'nodev', # keep ubuntu bootloader?
+            ('fileSystems', '/'): { 'device': '/dev/vda1', 'fsType': 'ext4'},
+            ('users', 'extraUsers', 'root', 'openssh', 'authorizedKeys', 'keys'): [self.depl.active_resources.get('ssh-key').public_key],
+        })
+
+    def get_ssh_private_key_file(self):
+        return self.write_ssh_private_key(self.depl.active_resources.get('ssh-key').private_key)
+
+    def create_after(self, resources, defn):
+        # make sure the ssh key exists before we do anything else
+        return {
+            r for r in resources if
+            isinstance(r, nixops.resources.ssh_keypair.SSHKeyPairState)
+        }
+
+    def get_auth_token(self):
+        return os.environ.get('DIGITAL_OCEAN_AUTH_TOKEN', self.auth_token)
+
+    def destroy(self, wipe=False):
+        self.log("destroying droplet {}".format(self.droplet_id))
+        try:
+            droplet = digitalocean.Droplet(id=self.droplet_id, token=self.get_auth_token())
+            droplet.destroy()
+        except digitalocean.baseapi.NotFoundError:
+            self.log("droplet not found - assuming it's been destroyed already")
+        self.public_ipv4 = None
+        self.droplet_id = None
+
+    def create(self, defn, check, allow_reboot, allow_recreate):
+        ssh_key = self.depl.active_resources.get('ssh-key')
+        if ssh_key is None:
+            raise Exception('Please specify a ssh-key resource (resources.sshKeyPairs.ssh-key = {}).')
+
+        if self.droplet_id is not None:
```

Comment:
(nitpick) log that it's a noop due to already existing resource. Helpful for debugging.